stdlib/REPL/src/REPL.jl

```diff
@@ -776,11 +776,46 @@ find_hist_file() = get(ENV, "JULIA_HISTORY",
 
 backend(r::AbstractREPL) = r.backendref
 
+# Allows an external package to add hooks into the code loading.
+# The hook should take a Vector{Symbol} of package names and
+# return true if all packages could be installed, false if not
+# to e.g. install packages on demand
+const install_packages_hooks = Any[]
+
 function eval_with_backend(ast, backend::REPLBackendRef)
+    if !isempty(install_packages_hooks)
+        check_for_missing_packages_and_run_hooks(ast)
+    end
     put!(backend.repl_channel, (ast, 1))
     return take!(backend.response_channel) # (val, iserr)
 end
 
+function check_for_missing_packages_and_run_hooks(ast)
+    mods = modules_to_be_loaded(ast)
+    filter!(mod -> isnothing(Base.identify_package(String(mod))), mods) # keep missing modules
+    if !isempty(mods)
+        for f in install_packages_hooks
+            f(mods) && return
+        end
+    end
+end
+
+function modules_to_be_loaded(ast, mods = Symbol[])
+    if ast.head in [:using, :import]
+        for arg in ast.args
+            if first(arg.args) isa Symbol # i.e. `Foo`
+                push!(mods, first(arg.args))
+            else # i.e. `Foo: bar`
+                push!(mods, first(first(arg.args).args))
```

Comment:
I think this might need to be made more robust. I don't think this works for something like `import A.b as c` yet. Does this currently know to properly ignore imports like `using .A`?

Reply:
I didn't think about either of those. I'll investigate and add tests

Reply:
Fixed the `using .A` case, and the other worked already. Added tests. Are there any other forms to watch out for?